arch/xtensa/soc/esp32/esp32-mp.c

```diff
@@ -0,0 +1,190 @@
+#include <zephyr.h>
+#include <kernel_structs.h>
+#include <soc.h>
+
+#define _REG(base, off) (*(volatile u32_t *)((base) + (off)))
+
+#define RTC_CNTL_BASE             0x3ff48000
+#define RTC_CNTL_OPTIONS0     _REG(RTC_CNTL_BASE, 0x0)
+#define RTC_CNTL_SW_CPU_STALL _REG(RTC_CNTL_BASE, 0xac)
+
+#define DPORT_BASE                 0x3ff00000
+#define DPORT_APPCPU_CTRL_A    _REG(DPORT_BASE, 0x02C)
+#define DPORT_APPCPU_CTRL_B    _REG(DPORT_BASE, 0x030)
+#define DPORT_APPCPU_CTRL_C    _REG(DPORT_BASE, 0x034)
+
+/* Two fields with same naming conventions live in two different
+ * registers and have different widths...
+ */
+#define RTC_CNTL_SW_STALL_APPCPU_C0  (0x03 << 0)  /* reg: OPTIONS0 */
+#define RTC_CNTL_SW_STALL_APPCPU_C1  (0x3f << 20) /* reg: SW_CPU_STALL */
+
+#define DPORT_APPCPU_CLKGATE_EN  BIT(0)
+#define DPORT_APPCPU_RUNSTALL    BIT(0)
+#define DPORT_APPCPU_RESETTING   BIT(0)
+
+/* These calls are ROM-resident and have fixed addresses. No, I don't
+ * know how they plan on updating these portably either.
+ */
+typedef void (*esp32rom_call_t)(int);
+static const esp32rom_call_t esp32rom_Cache_Flush = (void*)0x40009a14;
```

Comment:
Before merging, please update linker.ld for esp32, and add the prototypes for these functions in arch/xtensa/soc/esp32/soc.h.